i6_models/parts/fsa.py

```diff
@@ -0,0 +1,136 @@
+__all__ = ["TorchFsaBuilder", "WeightedFsa"]
+
+from functools import reduce
+from typing import Iterable, NamedTuple, Tuple, TypeVar
+
+import numpy as np
+import torch
+
+TWeightedFsa = TypeVar("TWeightedFsa", bound="WeightedFsa")
+
+
+class WeightedFsa(NamedTuple):
+    """
+    Convenience class that represents an FSA. It supports scaling the weights of the
+    fsa by simple left-multiplication and moving the tensors to a different device.
+    It can simply be passed to `i6_native_ops.fbw.fbw_loss` and `i6_native_ops.fast_viterbi.align_viterbi`.
+    :param num_states: the total number of all states S
+    :param edges: a [4, E] tensor of edges with number of edges E and where each column is an edge
+        consisting of from-state, to-state, emission idx and the index of the sequence it belongs to
+    :param weights: a [E,] tensor of weights for each edge scaled by the tdp_scale
+    :param start_end_states: a [N, 2] tensor of start and end states for each of the N sequences
+    """
+
+    num_states: torch.IntTensor
+    edges: torch.IntTensor
+    weights: torch.FloatTensor
+    start_end_states: torch.IntTensor
+
+    def __mul__(self: TWeightedFsa, scale: float) -> TWeightedFsa:
+        """Multiply the weights, i.e. the third element, with a scale."""
+        return WeightedFsa._make(tensor * scale if i == 2 else tensor for i, tensor in enumerate(self))
```

Comment:
I think having this explicit is better:
```suggestion
        return WeightedFsa(
            self.num_states,
            self.edges,
            self.weights * scale,
            self.start_end_states,
        )
```